server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -377,4 +384,17 @@ public TaskResult result(DiscoveryNode node, ActionResponse response) throws IOE
             throw new IllegalStateException("response has to implement ToXContent to be able to store the results");
         }
     }
+
+    /**
+     * Awaits for stopThreadResourceTracking to be called for all threads of the current task.
+     * @throws InterruptedException if thread interrupted while waiting
+     * @throws TimeoutException if timed out while waiting
+     */
+    public void awaitResourceTrackingThreadsCompletion() throws InterruptedException, TimeoutException {
+        resourceTrackingThreadsBarrier.awaitAdvanceInterruptibly(
+            resourceTrackingThreadsBarrier.arriveAndDeregister(),
+            10L,
+            TimeUnit.SECONDS
+        );
```

Comment:
What thread pool is blocked on await. Is there a mechanism we could make it wait-free?

Reply:
Await is only performed during Task unregistration; stopping resource tracking for a task's thread remains await-free.

Reply:
Another alternative could be to modify the code flow – calling unregister only when the last thread of the task has completed in order to make it await-free. I need to explore more on its possibility though.